pkg/rpcbackend/backend.go

```diff
@@ -141,11 +168,137 @@ func (rc *RPCClient) CallRPC(ctx context.Context, result interface{}, method str
 	return nil
 }
 
-// SyncRequest sends an individual RPC request to the backend (always over HTTP currently),
-// and waits synchronously for the response, or an error.
-//
-// In all return paths *including error paths* the RPCResponse is populated
-// so the caller has an RPC structure to send back to the front-end caller.
+type batchRequest struct {
+	rpcReq *RPCRequest
+	rpcRes chan *RPCResponse
+	rpcErr chan error
+}
+
+func (rc *RPCClient) startBatcher(ctx context.Context, batchDelay time.Duration, batchSize int) {
+	requestQueue := make(chan *batchRequest)
+
+	go func() {
+		ticker := time.NewTicker(batchDelay)
+		defer ticker.Stop()
+
+		var batch []*batchRequest
+
+		for {
+			select {
+			case req := <-requestQueue:
+				batch = append(batch, req)
+				if len(batch) >= batchSize {
+					rc.sendBatch(ctx, batch)
+					batch = nil
+				}
+			case <-ticker.C:
+				if len(batch) > 0 {
+					rc.sendBatch(ctx, batch)
+					batch = nil
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	rc.requestBatchQueue = requestQueue
+}
+
+func (rc *RPCClient) sendBatch(ctx context.Context, batch []*batchRequest) {
+	select {
+	case rc.requestBatchWorkerSlots <- true:
```

Comment:
I'm understanding your algorithm design is the other way round to the batch workers we have already (like the writer one I linked to above). You regularly tick to see if you can assemble a batch, then kick off a worker with that batch whether it's full or not.

Rather than having the workers pulling the requests.

This is sound logic as I think you will optimize for density 👍 

... but I still think you need the assembler go routine to try and fill a batch using a timer that starts with the first entry in the batch, rather than dispatching whatever happens on a tick.

At the two extreme ends:
- With a batch size of zero/near-zero the above logic will use a lot of CPU spinning on an empty set
- With a long batch size like 500ms, the chance to fill a batch after the first request could vary wildly depending on the proximity of the tick to the first request

Reply:
that's a great catch, will make sure the ticker starts when the first request is added to the batch